bakerytest/bakerytest.go

```diff
@@ -127,3 +130,131 @@ func (d *Discharger) PublicKeyForLocation(loc string) (*bakery.PublicKey, error)
 	}
 	return nil, bakery.ErrNotFound
 }
+
+type discharge struct {
+	cavId string
+	c     chan error
+}
+
+// LoginDischarger is a Discharger that always requires interraction to
+// complete the discharge.
+type LoginDischarger struct {
+	Discharger
+	Mux *http.ServeMux
+
+	// mu protects the following fields.
+	mu      sync.Mutex
+	waiting map[string]discharge
+	id      int
+}
+
+// NewLoginDischarger returns a new LoginDischarger which uses
+// loginHandler to handle the login interraction. The login is completed
+// by calling FinishLogin.
+//
+// If locator is non-nil, it will be used to find public keys
+// for any third party caveats returned by the checker.
+//
+// Calling this function has the side-effect of setting
+// InsecureSkipVerify in http.DefaultTransport.TLSClientConfig
+// until all the dischargers are closed.
+func NewLoginDischarger(locator bakery.PublicKeyLocator, loginHandler http.Handler) *LoginDischarger {
```

Comment:
Given the terminology we use elsewhere in the bakery, I wonder whether this should be "NewInteractiveDischarger", and loginHandler could be visitHandler.

I also wonder if, instead of implementing http.Handler, loginHandler could be:

    func(http.ResponseWriter, *http.Request) error

and we could call FinishLogin automatically when it returns (and unexport FinishLogin).